### PR TITLE
Md2 - hash implementation added according to RFC 1319

### LINKED ISSUE #3582: Support for MD2 hashing function
Despite being outdated for a long time, it's still possible to see it used in some old binaries and firmwares. OpenSSL can't be used for it since it is deprecated and will be removed relatively soon. It has to be implemented independently.

See `phl` command and these files:
- `librz/hash/p/*`
- `librz/hash/algorithms/*`

### PR DESCRIPTION
**Pull request by rockrid3r**
Fixes #3582 

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've documented or updated the documentation of every function and struct this PR changes. If not so I've explained why.
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [x] I've updated the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Added implementation of md2-hash by using the work of **deroad**(couldn't find on github) and [RFC-1319.](https://datatracker.ietf.org/doc/html/rfc1319)

**Test plan**

The following is the test suite from RFC 1319:
```
MD2 ("") = 8350e5a3e24c153df2275c9f80692773
MD2 ("a") = 32ec01ec4a6dac72c0ab96fb34c0b5d1
MD2 ("abc") = da853b0d3f88d99b30283a69e6ded6bb
MD2 ("message digest") = ab4f496bfb2a530b219ff33031fe06b0
MD2 ("abcdefghijklmnopqrstuvwxyz") = 4e8ddff3650292ab5a4108c3aa47940b
MD2 ("ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789") =
da33def2a42df13975352846c30338cd
MD2 ("123456789012345678901234567890123456789012345678901234567890123456
78901234567890") = d5976f79d83d3a0dc9806c3c66f3efd8
```
Can be tested in bash:
```
rz-hash -a md2 -s "abc" 
```
output:
```
string: 0x00000000-0x00000003 md2: da853b0d3f88d99b30283a69e6ded6bb
```
as required.

Also added test to `test_hash.c`.

### CODE REVIEW

File: librz/hash/algorithms/md2/md2.c [hunk 1]
```diff
@@ -0,0 +1,131 @@
+#include "md2.h"
```
Comment (XVilka) [inline librz/hash/algorithms/md2/md2.c:4]: Please add SPDX headers

File: librz/hash/p/algo_md2.c [hunk 2]
```diff
@@ -0,0 +1,83 @@
+#include <rz_hash.h>
```
Comment (XVilka) [inline librz/hash/p/algo_md2.c:4]: SPDX

**Review by XVilka — CHANGES_REQUESTED** [review]
You also need to update tests with the new output, e.g.
```
[XX] C:\projects\rizin\test\db\tools\rz_hash rz-hash -c with all hash algorithms
ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.6.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '!rz-hash -a all -s "admin" -c 21232f297a57a5a743894a0e4a801fc3' =
-- stdout
--- expected
+++ actual
@@ -1,3 +1,4 @@
+string: 0x00000000-0x00000005 md2: computed hash doesn't match the expected one
 string: 0x00000000-0x00000005 md4: computed hash doesn't match the expected one
 string: 0x00000000-0x00000005 md5: computed hash matches the expected one
 string: 0x00000000-0x00000005 sha1: computed hash doesn't match the expected one
[XX] C:\projects\rizin\test\db\tools\rz_hash rz-hash -L
```
See more at https://ci.appveyor.com/project/rizinorg/rizin/builds/47328795/job/6rx33say5x6xebam

**Comment by rockrid3r** [comment]
> You also need to update tests with the new output, e.g.
> 
> ```
> [XX] C:\projects\rizin\test\db\tools\rz_hash rz-hash -c with all hash algorithms
> ANSICON=1 RZ_NOPLUGINS=1 C:\projects\rizin\rizin-vs2022_64-v0.6.0\bin\rizin.exe -escr.utf8=0 -escr.color=0 -escr.interactive=0 -eflirt.sigdb.load.system=false -eflirt.sigdb.load.home=false -N -Qc '!rz-hash -a all -s "admin" -c 21232f297a57a5a743894a0e4a801fc3' =
> -- stdout
> --- expected
> +++ actual
> @@ -1,3 +1,4 @@
> +string: 0x00000000-0x00000005 md2: computed hash doesn't match the expected one
>  string: 0x00000000-0x00000005 md4: computed hash doesn't match the expected one
>  string: 0x00000000-0x00000005 md5: computed hash matches the expected one
>  string: 0x00000000-0x00000005 sha1: computed hash doesn't match the expected one
> [XX] C:\projects\rizin\test\db\tools\rz_hash rz-hash -L
> ```
> 
> See more at https://ci.appveyor.com/project/rizinorg/rizin/builds/47328795/job/6rx33say5x6xebam


You mean I should manually change the file `test/db/cmd/cmd_list` and `test/db/tools/rz_hash`?

File: librz/hash/meson.build [hunk 3]
```diff
@@ -112,6 +114,7 @@ if sys_openssl.found()
   dependencies += [sys_openssl]
 else
   rz_hash_sources += [
+    'algorithms/md2/md2.c',
```
Comment (wargio) [inline librz/hash/meson.build]: this should be probably available regardless of openssl.

**Review by wargio — CHANGES_REQUESTED** [review]
Please add a test under db/tools/rz_hash

File: librz/hash/meson.build (hunk 3 above)
Comment (wargio) [reply librz/hash/meson.build]: this is because you haven't added the openssl implementation

File: librz/hash/algorithms/md2/md2.c [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+
+	memcpy(buf, state, state_size);
+	memcpy(buf + state_size, block, block_size);
+
+	for (ut64 j = 0; j < block_size; ++j) {
+		buf[j + block_size + state_size] = block[j] ^ state[j];
+	}
+
+	ut64 t = 0;
+	for (ut64 j = 0; j < 18; ++j) {
```
Comment (wargio) [inline librz/hash/algorithms/md2/md2.c]: change that `18` into a define so we know is because it has 18 rounds.

**Review by XVilka — CHANGES_REQUESTED** [review]
1. Please rebase against latest `dev`
2. There is one unused var:
```
 FAILED: librz/hash/librz_hash.0.6.dylib.p/algorithms_md2_md2.c.o 
clang -Ilibrz/hash/librz_hash.0.6.dylib.p -I. -I.. -Ilibrz -I../librz -Ilibrz/include -I../librz/include -Ilibrz/util/sdb/src -I../librz/util/sdb/src -Ilibrz/util/sdb -I../librz/util/sdb -Isubprojects/xxhash -I../subprojects/xxhash -I../subprojects/blake3/c -fcolor-diagnostics -Wall -Winvalid-pch -Werror -O3 -DRZ_PLUGIN_INCORE=1 --std=gnu99 -Werror=sizeof-pointer-memaccess -fvisibility=hidden -MD -MQ librz/hash/librz_hash.0.6.dylib.p/algorithms_md2_md2.c.o -MF librz/hash/librz_hash.0.6.dylib.p/algorithms_md2_md2.c.o.d -o librz/hash/librz_hash.0.6.dylib.p/algorithms_md2_md2.c.o -c ../librz/hash/algorithms/md2/md2.c
../librz/hash/algorithms/md2/md2.c:108:13: error: variable 'n_called' set but not used [-Werror,-Wunused-but-set-variable]
        static int n_called = 0;
                   ^
1 error generated.
```

**Review by wargio — APPROVED** [review]
Awesome job!

**Review by XVilka — APPROVED** [review]

**Comment by XVilka** [comment]
Thank you!

**Pull request merged**